### PR TITLE
Fix Issue 23814 - [Codegen] Calling member function of extern(C++) cl…

### PR DESCRIPTION
**Pull request by naydef**
…ass with multiple inheritance doesn't preserve the EBX register in some cases

Remade.
Also not much idea of what I'm doing.
~~Not sure how to test this.~~

### CODE REVIEW

**Comment by RazvanN7** [comment]
As stated in the issue, I cannot reproduce this for neither 32 or 64 bit.

32 bit output:

```asm
00000000 <_ZN9MainClass5func4Ev>:
   0:   55                      push   ebp
   1:   8b ec                   mov    ebp,esp
   3:   83 ec 08                sub    esp,0x8
   6:   e8 00 00 00 00          call   b <_ZN9MainClass5func4Ev+0xb>
   b:   58                      pop    eax
   c:   05 02 00 00 00          add    eax,0x2
  11:   89 45 fc                mov    DWORD PTR [ebp-0x4],eax
  14:   8b 4d fc                mov    ecx,DWORD PTR [ebp-0x4]
  17:   8d 81 40 00 00 00       lea    eax,[ecx+0x40]
  1d:   c9                      leave  
  1e:   c3                      ret   
```

Does not use ebx.

64 bit output:

```asm
0000000000000000 <_ZN9MainClass5func4Ev>:
   0:   48 8d 05 00 00 00 00    lea    rax,[rip+0x0]        # 7 <_ZN9MainClass5func4Ev+0x7>
   7:   c3                      ret
```

Does not use rbx.

I don't think this issue is valid.

**Comment by naydef** [comment]
I also replied in the issue report:

I'm using DMD64 D Compiler v2.102.2 on Linux. Compiling with: "dmd app.d -m32"

I see you check func4(judging by name of the disassembly "_ZN9MainClass5func4Ev"), while my example code calls func1.
Use the second example code. You can use a debugger and break where func1 is called in _Dmain, step into the function and you'll see the usage of EBX register in a function called _THUNK0, which at the end jumps to _ZN9MainClass5func1Ev.

This issue has actually caused me trouble when interoperating with C++ code on Linux, it doesn't seem to exist on Windows.

**Review by Geod24 — DISMISSED** [review]
1) This needs a test (which can then be used to repro the issue @RazvanN7 )
2) If the code is wrong, remove it, don't make the `if` always false.

**Comment by naydef** [comment]
ping (Requested changes are applied)

**Comment by RazvanN7** [comment]
cc @WalterBright

**Comment by naydef** [comment]
@RazvanN7
Can this pull request be reviewed? It's blocking my project on Linux. I'll prefer It if the pull request get outright dismissed than rotting here for ages.

**Comment by dkorpel** [comment]
Thanks for the ping, I've force pushed to restart the test suite and emailed Walter to request his review, since he wrote the backend.

**Comment by naydef** [comment]
@dkorpel
Can test suite be restarted (failure seems unrelated)?

**Comment by dkorpel** [comment]
I can rebase again, but @RazvanN7 has access to buildkite directly

**Comment by dkorpel** [comment]
Or merge despite the failure

**Comment by RazvanN7** [comment]
I restarted the buildkite testing pipeline. Adding the `72h no objection -> merge` label.

**Review by WalterBright — CHANGES_REQUESTED** [review]

File: compiler/src/dmd/backend/cod3.d [hunk 1]
```diff
@@ -4843,18 +4843,6 @@ void cod3_thunk(Symbol *sthunk,Symbol *sfunc,uint p,tym_t thisty,
...
-            localgot = null;                // no local variables
-            CodeBuilder cdbgot; cdbgot.ctor();
-            load_localgot(cdbgot);          // load GOT in EBX
-            code *c1 = cdbgot.finish();
-            if (c1)
-            {
-                assignaddrc(c1);
-                cdb.append(c1);
-            }
-        }
```
[the remark this answers is not in the document]
Comment (naydef) [reply compiler/src/dmd/backend/cod3.d:4857]: I don't have much clue how to do that (not familiar with compiler backends).

[the remark this answers is not in the document]
Comment (WalterBright) [reply compiler/src/dmd/backend/cod3.d:4857]: I wrote a more explanatory comment [in the bu](https://issues.dlang.org/show_bug.cgi?id=23814#c6)

[the remark this answers is not in the document]
Comment (WalterBright) [reply compiler/src/dmd/backend/cod3.d:4857]: ... in the bug report

**Comment by adamdruppe** [comment]
Merged in OpenD.

**Comment by RazvanN7** [comment]
Since @WalterBright opposes the current implementation I see no point in keeping this PR open. If anyone wants to try and fix the issue there is a link from the bug report to the this PR.

**Comment by RazvanN7** [comment]
@naydef Thanks for your work and the time you put into this. I'm sorry that this is not the outcome you desired, however, maybe this will prompt @WalterBright to fix this issue.

**Pull request closed without merging**